Apiserver: wait for Etcd to become available on startup

stu-gott opened the pull request:
fixes #37704

stu-gott committed:
Apiserver: wait for etcd to become available on startup

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39821)
<!-- Reviewable:end -->

stu-gott commented:
I signed the CLA

xiangpengzhao commented on cmd/kube-apiserver/apiserver.go:
```diff
@@ -47,6 +48,11 @@ func main() {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
+	}
+
```
How about putting this into app.Run in order to keep the main func clean ?

nikhiljindal commented:
Assigning to @sttts

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+)
+
+var retryLimit int = 6
+var retryInterval time.Duration = 10
```
10 nanoseconds is a pretty small interval ;)

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+type realTimer struct{}
+func (realTimer) Sleep(d time.Duration) {time.Sleep(d)}
+
+var timer clock = new(realTimer)
```
var _ clock = ...

stu-gott replied:
I don't understand the purpose of using underscore here. timer is actually used. Perhaps I'm unclear on the construct, but the point here is to allow mocking out sleeping for 10 second intervals (so the test suite won't take 4 minutes while literally sleeping). Is there a better way to go about this?

sttts replied:
Global variables are an anti-pattern in go because they are hard to reason about and lead to synchronization issues. Can't one inject this somehow into the wait logic?

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+
+type realEtcdConnection struct {}
+
+var etcdConnection connection = new(realEtcdConnection)
```
_

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+var timer clock = new(realTimer)
+
+type connection interface {
+	checkConnection(string) bool
```
address string

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+
+var etcdConnection connection = new(realEtcdConnection)
+
+func (realEtcdConnection) checkConnection(connString string) bool {
```
s/connString/address/

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+
+		done := etcdConnection.checkConnection(connUrl.Host)
+		retries := 0
+		for (!done) && (retries < retryLimit) {
```
no need for the parenthesis

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+			timer.Sleep(retryInterval)
+			done = etcdConnection.checkConnection(connUrl.Host)
+		}
+		if retries >= retryLimit {
```
the last `checkConnection` in the loop could have been successful. Better check for `!done`

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+func WaitForEtcd(serverList []string) error {
+	for _, connString := range serverList {
```
I would expect a parallel wait for all servers. In case of a network problem to one etcd server, we retry here for quite some time and also time out. Shouldn't we be happy if we connect to one etcd which claims to have quorum?

stu-gott replied:
Thank you for clearing that up! I was of the mis-impression that all etcd servers needed to be reached (e.g., like a Redis or memcached setup)

sttts commented on cmd/kube-apiserver/app/preflight/checks_test.go:
```diff
@@ -0,0 +1,97 @@
+	return (checkCallCount >= (timesToFail + 1))
+}
+
+var checkCallCount int
```
global variables in tests? 😱

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -95,6 +96,10 @@ func Run(s *options.ServerRunOptions) error {
 	if err := s.CloudProvider.DefaultExternalHost(s.GenericServerRunOptions); err != nil {
 		return fmt.Errorf("error setting the external host value: %v", err)
 	}
+	if err := preflight.RunApiserverChecks(s); err != nil {
```
If we want this here, then after all options are validated, i.e. around line 120.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+package preflight
+
+import (
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
```
convention: go packages first, empty line, other packages, empty line, kube packages

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+	"errors"
+)
+
+var retryLimit int = 6
```
all these are constants.

xiangpengzhao commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+
+func (realEtcdConnection) checkConnection(connString string) bool {
+	if conn, err := net.Dial("tcp", connString); err != nil {
+		conn.Close()
```
no need to Close here

stu-gott replied:
actually, I meant to have used 

if ....; err == nil {

xiangpengzhao commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+func (realEtcdConnection) checkConnection(connString string) bool {
+	if conn, err := net.Dial("tcp", connString); err != nil {
+		conn.Close()
+		return true
```
return false here and return true else. Maybe I misunderstanding the logic.

xiangpengzhao commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,79 @@
+	if conn, err := net.Dial("tcp", connString); err != nil {
+		conn.Close()
+		return true
+	}
```
`defer conn.Close()` after the brace because we just want to check the connection.

xiangpengzhao reviewed (commented):
agree with @sttts 's comments.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,96 @@
+
+	// overall timeout
+	go func() {
+		timer.Sleep(retryInterval * retryLimit)
```
There is https://golang.org/pkg/time/#After for the same thing.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,96 @@
+
+	select {
+	case <-foundEtcd:
+		{return nil}
```
no need for {}

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,96 @@
+	case <-foundEtcd:
+		{return nil}
+	case <- timeout:
+		{return errors.New("Unable to reach any Etcd server")}
```
errors are lower-case because they will be appended to each other like `something happened: %v`.

I believe etcd is lower-case usually.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,96 @@
+)
+
+const retryLimit = 6
+const retryInterval = 10 * time.Second
```
in fact, we can reduce this to something much lower like a second. 10 seconds is quite a lot.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,96 @@
+
+var etcd connection = new(etcdConnection)
+
+func (etcdConnection) checkConnection(address string) bool {
```
better use a name which says something about the semantics, like serverReachable.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,96 @@
+	retries := 0
+	for retries < retryLimit {
+		retries += 1
+		timer.Sleep(retryInterval)
```
Better do the sleep after the first check. Otherwise, we wait for no reason.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,96 @@
+	// Attempt to reach every Etcd server in goroutines
+	for _, serverURI := range serverList {
+		connUrl, _ := url.Parse(serverURI)
+		go checkEtcdServer(connUrl.Host, foundEtcd)
```
can we have a `stop` channel which is closed here from outside as soon as the first etcd answers? Otherwise, the polling loops will keep running.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,96 @@
+
+type etcdConnection struct {}
+
+var etcd connection = new(etcdConnection)
```
It's much more go idiomatic to turn `func (etcdConnection) checkConnection(address string) bool` into a `Checker` interface and inject that into `checkEtcdServer`. In the test you can use your mock this way. No need for the global variable.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,96 @@
+	return false
+}
+
+func checkEtcdServer(address string, foundEtcd chan struct{})  {
```
Inject your clock here as a parameter.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,96 @@
+		timer.Sleep(retryInterval)
+		if etcd.checkConnection(address) {
+			select {
+			case foundEtcd <- struct{}{}:
```
Why the select?

stu-gott replied:
The tl;dr is I intend for this to be non-blocking.

My impression of what will happen here is that if the buffer for the foundEtcd channel is filled (it has a buffer with 1 slot), any goroutine which attempts to send will block here, which would keep them open indefinitely.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,96 @@
+
+	// Attempt to reach every Etcd server in goroutines
+	for _, serverURI := range serverList {
+		connUrl, _ := url.Parse(serverURI)
```
The error should be checked here.

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -115,6 +116,9 @@ func Run(s *options.ServerRunOptions) error {
 		return fmt.Errorf("failed to configure authentication: %s", err)
 	}
+	if err := preflight.RunApiserverChecks(s); err != nil {
+		return fmt.Errorf("error starting apiserver: %v\n", err)
```
no newline in errors

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,102 @@
+		}
+		if etcd.serverReachable(address) {
+			select {
+			case foundEtcd <- struct{}{}:
```
case <- stop: return

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+func WaitForEtcd(serverList []string, timer clock, etcd connection) error {
+	foundEtcd := make(chan struct{}, 1)
```
all but the first go routines will block on foundEtcd.

stu-gott replied:
Actually, no go routines will block. In checkEtcdServer, the attempt to send on foundEtcd is in a select so is asynchronous. each routine (that would have blocked) will attempt to send, fail, and choose the default case (which is a no-op). Each one will then sleep and stop.

sttts replied:
Right.

sttts commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,102 @@
+import (
+	"net"
+	"net/url"
+	"time"
```
order is wrong

sttts commented:
/lgtm

Assigning to @deads2k for approval.

deads2k commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,101 @@
+	After(time.Duration) <-chan time.Time
+}
+
+type connectionTimer struct{}
```
no need for this.  Just use the util/clock.  apiservers already depend on client-go.

deads2k commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,101 @@
+	}
+}
+
+func RunApiserverChecks(s *options.ServerRunOptions) error {
```
`RunAPIServerChecks`

deads2k replied:
Actually, narrow name for a narrow case.  `WaitForAvailableEtcd`. server list and stopchannel as args.

deads2k commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,101 @@
+
+func checkEtcdServer(address string, foundEtcd chan struct{}, stop chan struct{}, timer clock, etcd connection) {
+	retries := 0
+	for retries < retryLimit {
```
this looks like `util/wait.PollImmediate`

deads2k replied:
Or `PollUntil` since its got the stop.

deads2k commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,101 @@
+	}
+}
+
+func WaitForEtcd(serverList []string, timer clock, etcd connection) error {
```
Needs godoc.  I gather that you're waiting until at least one is available?

deads2k commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -115,6 +116,9 @@ func Run(s *options.ServerRunOptions) error {
 	if err = s.Authentication.Apply(genericConfig); err != nil {
 		return fmt.Errorf("failed to configure authentication: %s", err)
 	}
+	if err := preflight.RunApiserverChecks(s); err != nil {
```
Alright, now making it all the way to usage, I think I'd like `utilwait.PollImmediate(interval, duration, conditionfunc)` where you provide the condition func for testing etcd servers as a function on an struct.


```go
err := utilwait.PollImmediate(100 milliseconds, 10 seconds, NewEtcdCheck(serverlist).Condition)
if err != nil{
    return err
}
```

stu-gott replied:
It's not clear to me how this will make the resulting code any cleaner--or even work. WaitForAvailableEtcd is waiting on multiple go routines simultaneously, and signalling for all of them to stop when a verdict has been reached. Because utilwait.PollImmediate is polling one condition in one thread, I'm not sure what we would gain from it here. Could you please elaborate on how you picture it being used here?

deads2k commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -115,6 +116,9 @@ func Run(s *options.ServerRunOptions) error {
 	if err = s.Authentication.Apply(genericConfig); err != nil {
 		return fmt.Errorf("failed to configure authentication: %s", err)
 	}
+	if err := preflight.RunAPIServerChecks(s); err != nil {
```
Alright, now making it all the way to usage, I think I'd like `utilwait.PollImmediate(interval, duration, conditionfunc)` where you provide the condition func for testing etcd servers as a function on an struct.


```go
err := utilwait.PollImmediate(100 milliseconds, 10 seconds, NewEtcdCheck(serverlist).Condition)
if err != nil{
    return err
}
```

> It's not clear to me how this will make the resulting code any cleaner--or even work. WaitForAvailableEtcd is waiting on multiple go routines simultaneously, and signalling for all of them to stop when a verdict has been reached. Because utilwait.PollImmediate is polling one condition in one thread, I'm not sure what we would gain from it here. Could you please elaborate on how you picture it being used here?

Throughout the rest of the codebase (more or less), we use `wait.<something>` to wait on conditions.  Since the retry is done separately, there's no retry logic in your "check for etcd logic".  Since you're only trying once per attempt, there's no reason to have the stop channel (a bounded number of extra gofuncs just waiting to fail or succeed doesn't much matter), so your logic above gets to focus on "is an etcd up".

deads2k replied:
`utilwait.PollImmediate(retryInterval, retryLimit, etcd.checkEtcdServers)` here?  

`utilwait.PollImmediate(retryInterval, retryLimit, etcd.checkEtcdServers(server list))`.  Since you can return an error from the condition, you could parse inside.  Multiple parses, but for this case it doesn't matter.

deads2k commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+func (etcdConnection) serverReachable(address string) bool {
+	if conn, err := net.Dial("tcp", address); err == nil {
```
does dial allow a timeout to be set?  If not, we can add it later.

deads2k commented:
looks nice.

smarterclayton commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -21,67 +21,46 @@ import (
 
-const retryLimit = 60
-const retryInterval = 1 * time.Second
+const RetryLimit = 60
```
Is there a reason for these constants to be public?  They're not tied to this package, and don't seem to provide much value for others.

stu-gott replied:
They're now being used in server.go which is outside this package.

smarterclayton replied:
Then they should be in server.go and private.

stu-gott commented:
@stts @deads2k @smarterclayton Is there general concurrence that this patch is good? As @mml pointed out, it would be good to work toward a configurable backoff strategy for crashlooping pods, but there's no harm in also implementing crash prevention in the kube-apiserver.

deads2k commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -115,6 +120,9 @@ func Run(s *options.ServerRunOptions) error {
 		return fmt.Errorf("failed to configure authentication: %s", err)
 	}
+	if err := utilwait.PollImmediate(etcdRetryInterval, etcdRetryLimit * etcdRetryInterval, preflight.EtcdConnection{ServerList: s.Etcd.StorageConfig.ServerList}.CheckEtcdServers); err != nil {
+		return fmt.Errorf("error starting apiserver: %v", err)
```
`error waiting for etcd connection: %v`

deads2k commented:
nit, squash, lgtm

deads2k commented:
> @stts @deads2k @smarterclayton Is there general concurrence that this patch is good? As @mml pointed out, it would be good to work toward a configurable backoff strategy for crashlooping pods, but there's no harm in also implementing crash prevention in the kube-apiserver.

Sorry, it didn't ping me after you pushed on the last set of changes.  I didn't notice.

deads2k commented:
@k8s-bot  ok to test

deads2k commented:
/lgtm
/approve
/release-note-none

deads2k commented:
`hack/update-bazel.sh` ought to make things right.

ncdc commented:
@stu-gott could you please rebase and rerun `hack/update-bazel.sh`?

stu-gott commented:
Whenever I run update-bazel.sh, it wrecks every BUILD file in my tree. (it looks like it removes every "//vendor:" dep). Since I've been unable to resolve that, this PR has remained in a broken state.

Any guidance on how to fix that would be greatly appreciated.

ncdc commented:
@stu-gott weird, I've never had that happen. @mikedanese is our bazel/gazel expert - any ideas?

ncdc commented:
I'll try pulling down your branch and updating bazel locally and report back.

ncdc commented:
@stu-gott I was able to pull your branch, rebase on top of master, and update bazel without issue. I'm trying to think what might cause issues... what is your GOPATH?

sttts commented:
@stu-gott can you post the diff of the BUILD files in a gist or push them here?

stu-gott commented:
Thanks for your help resolving my bazel/gazel issues @stts @ncdc

sttts commented:
@stu-gott what was it at the end?

stu-gott commented:
I re-created my $GOPATH from scratch. I assume there was something old and forgotten buried in there that was causing issues.

ncdc commented:
@k8s-bot node e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot gci gke e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot kops aws e2e test this
@k8s-bot non-cri node e2e test this
@k8s-bot unit test this
@k8s-bot verify test this

ncdc commented on cmd/kube-apiserver/app/preflight/checks.go:
```diff
@@ -0,0 +1,68 @@
+	for _, serverUri := range con.ServerList {
+		host, err := parseServerURI(serverUri)
+		if err != nil {
+			return false, err
```
Shouldn't this `continue` instead of returning early?

deads2k replied:
> Shouldn't this continue instead of returning early?

On a parse failure?  I'm ok failing early and hard on bad input.

ncdc replied:
Yeah that's fine now that I think about it more.

ncdc commented:
@k8s-bot cvm gke e2e test this

ncdc commented:
@smarterclayton @k8s-oncall this hopefully should mitigate the various test flakes where the RBAC roles fail to create and components/clients are denied when trying to talk to the apiserver

nikhiljindal commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -129,6 +134,9 @@ func Run(s *options.ServerRunOptions) error {
 	if err := s.Features.ApplyTo(genericConfig); err != nil {
 		return err
 	}
+	if err := utilwait.PollImmediate(etcdRetryInterval, etcdRetryLimit*etcdRetryInterval, preflight.EtcdConnection{ServerList: s.Etcd.StorageConfig.ServerList}.CheckEtcdServers); err != nil {
```
How about putting this in GenericApiServer? Maybe in GenericApiServer.PrepareRun or Run?
This doesnt see specific to kube-apiserver. federation-apiserver, for ex, should also have this?

ncdc replied:
Sounds reasonable

cc @deads2k @sttts

nikhiljindal commented:
I dont know how critical this PR is but I am inclined to remove LGTM label to give the author a chance to update the PR to move the code to genericapiserver.
@ncdc @deads2k @sttts Is it fine to remove the LGTM label?

ncdc commented:
I'm ok with that since #42186 should fix some of this.

On Mon, Feb 27, 2017 at 7:18 PM Nikhil Jindal <notifications@github.com>
wrote:

> I dont know how critical this PR is but I am inclined to remove LGTM label
> to give the author a chance to update the PR to move the code to
> genericapiserver.
> @ncdc <https://github.com/ncdc> @deads2k <https://github.com/deads2k>
> @sttts <https://github.com/sttts> Is it fine to remove the LGTM label?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/39821#issuecomment-282900779>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYvnAa4RKOzRUHotxE8ivtHKrmc6Iks5rg2fogaJpZM4LiMfx>
> .
>

deads2k commented:
> I dont know how critical this PR is but I am inclined to remove LGTM label to give the author a chance to update the PR to move the code to genericapiserver.
> @ncdc @deads2k @sttts Is it fine to remove the LGTM label?

Did we just miss the release for this?  I don't mind pushing some of these checks down, but this seems like letting perfect be the enemy of good.  Now we can't get either.

stu-gott commented:
I have unrelated deadlines so don't have the bandwidth to concentrate on this in the short term. If it's critical that this patch be moved to genericapiserver, can somebody else take ownership of that?

fabiand commented:
Agreeing with @deads2k 

There is a fix, which went through a few iterations - and considering that this was a beginner item, I think it would just be fair to merge it and find follow up fixes for the GenericApiServer. Just my 2ct.

ncdc commented:
re-adding lgtm since this is a bug fix and based on conversation with @idvoretskyi in slack

/lgtm

ncdc commented:
/lgtm

ncdc commented:
```
I0228 17:27:26.777] k8s.io/kubernetes/cmd/kube-apiserver/app/server.go:159: not enough arguments to return
```

looks like this is going to need a rebase & minor fix

stu-gott commented:
Rebased and changed line 159:

-               return fmt.Errorf("error waiting for etcd connection: %v", err)
+               return nil, nil, fmt.Errorf("error waiting for etcd connection: %v", err)

ncdc commented:
/lgtm

ncdc commented:
@k8s-bot ok to test

nikhiljindal commented:
Yes the intention was not to push this out of release. This had LGTM and approval labels before the code freeze deadline so we should be good.

I would like to reiterate that it is critical to fix bugs for all apiservers (wherever appropriate) instead of fixing kube-apiserver only. We want to keep federation and kube apiservers in sync. Fixes to only one of them makes it harder to find and debug issues in the other.

deads2k commented:
> I would like to reiterate that it is critical to fix bugs for all apiservers (wherever appropriate) instead of fixing kube-apiserver only. We want to keep federation and kube apiservers in sync. Fixes to only one of them makes it harder to find and debug issues in the other.

It's equally important to remember that the generic api server library is for *all* api servers, not just kube and federation.  There is a shared library just for them under `pkg/kubeapiserver`.  This particular condition check might be moved eventually, but since storage is configured by interface, not by pinning to etcd it cannot be an unconditional check like this.

nikhiljindal commented:
aah nice! Thanks for pointing out pkg/kubeapiserver @deads2k.  Yes all I want is common code (not necessarily in genericapiserver) :)

ethernetdan commented:
Removing PR from 1.6 milestone since code freeze has begun:
* If this is a bug fix, please label as `kind/bug` and reapply the milestone.
* If this is a feature for 1.6, an [exception](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) should be filed for it before 11:00PM PST on 3/2. Note that most exceptions to code freeze are not accepted.
* If this is a feature for 1.7, feel free to add the 1.7 milestone to the PR.

/cc @kubernetes/release-team @kubernetes/kubernetes-release-managers

ncdc commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot kops aws e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot cvm gke e2e test this
@k8s-bot gci gke e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot verify test this
@k8s-bot unit test this
@k8s-bot non-cri node e2e test this
@k8s-bot node e2e test this
